webapp/packages/plugin-data-grid/src/GrantManagementTable.tsx

```diff
@@ -0,0 +1,221 @@
+import { TableRowSelect } from './TableRowSelect.js';
+import classes from './GrantManagementTable.module.css';
+
+interface IColumn {
+  key: string;
+  label: TLocalizationToken;
+}
+
+const SELECT_COLUMN = { key: 'select', label: '' };
+const STATUS_COLUMN = { key: 'status', label: 'ui_granted' };
+
+const DEFAULT_COLUMNS: IColumn[] = [SELECT_COLUMN, STATUS_COLUMN];
+
+export interface IGrantManagementTableProps<T> {
+  items: T[];
+  columns: IColumn[];
+  getItemId: (item: T) => string;
+  isGranted: (item: T) => boolean;
+  isEdited: (item: T) => boolean;
+  onGrant: (ids: string[]) => void;
+  onRevoke: (ids: string[]) => void;
+  disabled?: boolean;
+  isManageable?: (item: T) => boolean;
+  isVisible?: (item: T, filter: string) => boolean;
+  getCell?: (item: T, colKey: string) => React.ReactNode;
+}
+
+export const GrantManagementTable = observer(function GrantManagementTable<T>({
+  columns,
+  items,
+  getItemId,
+  isGranted,
+  isEdited,
+  onGrant,
+  onRevoke,
+  disabled,
+  isManageable,
+  isVisible,
+  getCell,
+}: IGrantManagementTableProps<T>) {
+  const translate = useTranslate();
+  const styles = useS(classes);
+
+  const [filter, setFilter] = useState('');
+  const deferredFilter = useDeferredValue(filter);
+
+  const visibleItems = useMemo(() => {
+    if (isVisible) {
+      return items.filter(item => isVisible(item, deferredFilter));
+    }
+
+    return items;
+  }, [isVisible, items, deferredFilter]);
+
+  const keys = useMemo(() => {
+    const filtered = isManageable ? visibleItems.filter(isManageable) : visibleItems;
+    return filtered.map(getItemId);
+  }, [visibleItems, getItemId, isManageable]);
+
+  const selection = useTableSelection(keys);
+
+  function grant() {
+    const prev = visibleItems.filter(item => isGranted(item)).map(getItemId);
+    const granted = selection.list.filter(id => !prev.includes(id));
+
+    onGrant(granted);
+    selection.clear();
+  }
+
+  function revoke() {
+    const prev = visibleItems.filter(item => isGranted(item)).map(getItemId);
+    const revoked = selection.list.filter(id => prev.includes(id));
+
+    onRevoke(revoked);
+    selection.clear();
+  }
+
+  const _columns = useMemo(() => [...DEFAULT_COLUMNS, ...columns], [columns]);
+
+  function _getCell(rowIdx: number, colIdx: number) {
+    const row = visibleItems[rowIdx] as T;
+    const column = _columns[colIdx];
+
+    if (!row || !column) {
+      return null;
+    }
+
+    if (column.key === SELECT_COLUMN.key) {
+      return <TableRowSelect id={getItemId(row)} disabled={isManageable?.(row) === false} />;
+    }
+
+    if (column.key === STATUS_COLUMN.key) {
+      const granted = isGranted(row);
+      return (
+        <div
+          className={clsx('tw:m-auto tw:h-2 tw:w-2 tw:rounded-full tw:bg-[var(--theme-background)]', granted && 'tw:bg-[var(--theme-positive)]')}
+        />
+      );
+    }
+
+    return getCell?.(row, column.key) ?? getItemId(row);
```

Comment:
Probably, we don't want to fall back to `getItemId`. Imagine you added a new custom column and forgot to process it in `getCell`, but you will see a value from `getItemId` that will be confusing. I will suggest returning null or an empty string in that case case